File: Grind75/72-BasicCalculator/cpp/BasicCalculator.cpp
#include "BasicCalculator.h"
#include <string>

using namespace std;

int calc(const string &s, uint &i) {
  int result = 0;
  int num = 0;
  int sign = 1;
  while (i < s.size() && s[i] != ')') {
    if (isdigit(s[i])) {
      num *= 10;
      num += int(s[i] - '0');
    } else {
      if (s[i] == '(') {
        num = calc(s, ++i);
        result += sign * num;
        sign = 1;
        num = 0;
      } else if (s[i] == '+') {
        result += sign * num;
        sign = 1;
        num = 0;
      } else if (s[i] == '-') {
        result += sign * num;
        sign = -1;
        num = 0;
      }
    }
    i++;
  }
  result += sign * num;
  return result;
}

int Solution::calculate(string s) {
  uint index = 0;
  return calc(s, index);
}

File: Grind75/72-BasicCalculator/python/BasicCalculator.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Tuple
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def calculate(self, s: str) -> int:
    stack: List[Tuple[int, int]] = []
    sgn, num, sum = 1, 0, 0
    for c in s:
      match c:
        case d if c.isdigit():
          num = num * 10 + int(d)
        case '+':
          sum += sgn * num
          sgn, num = 1, 0
        case '-':
          sum += sgn * num
          sgn, num = -1, 0
        case '(':
          stack.append( (sgn, sum) )
          sgn, sum, num = 1, 0, 0
        case ')':
          sum += sgn * num
          beforeSgn, beforeSum = stack.pop()
          sum = beforeSum + beforeSgn * sum
          sgn, num = 1, 0
    sum += sgn * num
    return sum

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( "", 0 ),
    ( "1", 1 ),
    ( "(1)", 1 ),
    ( "1+2", 3 ),
    ( "1+ 2", 3 ),
    ( "1 +2", 3 ),
    ( "1 + 2", 3 ),
    ( "1 - 2", -1 ),
    ( "-1-2", -3 ),
    ( "1-2 + 3", 2 ),
    ( "-0-2", -2 ),
    ( "1 + 2 + 3", 6 ),
    ( "1 + 2 + (3 + 4)", 10 ),
    ( "1 + 2 - (3 + 4)", -4 ),
    ( "1 + 2 - (3 + (4 - 2))", -2 ),
  ]

  for s, answer in testCases:
    colorPrint(LIGHT_BLUE, quote(s), '=>', answer)
    result = soln.calculate(s)
    colorPrint(GREEN if result == answer else RED, ' -->', result)

